Add SPI basic test and test header file.

mprse opened the pull request:
### Description

Provide SPI test and test header file based on the new HAL API description provided here:
https://github.com/ithinuel/mbed-os/blob/rfc-spi-overhaul/hal/rfcs/0000-spi-overhaul.md

This PR can not be merged until PR which defines the new HAL API and adds at least one example implementation is merged. Probably then the test will have to be adapted to the final API version.

This test is a first part of entire SPI testing scheme which provides non-communication, full-automated test which will run on CI.

| PR | Description |
| --- | --- |
| RFC SPI: Reference Implementation. #8445 | New HAL API definition and example implementations for K66F anf NUCLEO_F429ZI |
| Add SPI test and test header file. #7976 | SPI Green Tea basic test |
| Add GreeTea SPI communication test. #8216 | Two boards Green Tea SPI communication test (closed) |
| Add Ice Tea SPI communication test. #8443 | Two boards  Ice Tea SPI communication test |
| Add one board Green Tea SPI Communication Test #8919  | One board Green Tea SPI communication test |

Please assign @ithinuel as a reviewer. 

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Breaking change

cmonr commented:
@ARMmbed/mbed-os-hal ?

fkjagodzinski commented on TESTS/mbed_hal/spi/main.cpp:
```diff
@@ -0,0 +1,689 @@
+
+        TEST_ASSERT(capabilities.minimum_frequency <= FREQ_200KHZ);
+        TEST_ASSERT(capabilities.maximum_frequency >= FREQ_2MHZ);
+        TEST_ASSERT_TRUE(capabilities.symbol_length & CAPABILITY_WORD_LENGTH_8);
```
It might be useful to comment this is actually defined behavior / requirement from the docs.

fkjagodzinski commented on TESTS/mbed_hal/spi/main.cpp:
```diff
@@ -0,0 +1,689 @@
+    spi_t spi_obj =
+    { 0 };
+    spi_capabilities_t capabilities =
+    { 0 };
```
These lines break for no reason.

fkjagodzinski commented on TESTS/mbed_hal/spi/main.cpp:
```diff
@@ -0,0 +1,689 @@
+    }
+
+    /* Indicate that test case has been successfully executed. */
+    TEST_ASSERT_TRUE(true);
```
This looks a little odd. I think this line is unnecessary for the test case to be valid, but isn't there anything to check during code execution?

fkjagodzinski commented on TESTS/mbed_hal/spi/main.cpp:
```diff
@@ -0,0 +1,689 @@
+
+        spi_format(&spi_obj, 8, SPI_MODE_IDLE_LOW_SAMPLE_FIRST_EDGE, SPI_BIT_ORDERING_MSB_FIRST);
+
+        real_freq = spi_frequency(&spi_obj, FREQ_200KHZ);
```
Looking at https://github.com/ARMmbed/mbed-os/blob/689890e1de06de8aa983e10316a4c5b861c63e75/hal/spi_api.h#L104 I can see there is no return value.

fkjagodzinski commented on TESTS/mbed_hal/spi/main.cpp:
```diff
@@ -0,0 +1,689 @@
+                    p_fill_sym = (void*) &fill_symbol_32;
+                }
+
+                const uint32_t sym_cnt = spi_transfer(&spi_obj, p_tx_buf, BUF_SIZE, p_rx_buf, BUF_SIZE, p_fill_sym);
```
You probably need to use `spi_master_block_write()` instead.

fkjagodzinski commented on TESTS/mbed_hal/spi/main.cpp:
```diff
@@ -0,0 +1,689 @@
+                    handler_called = false;
+                    expected_event.transfered = 0; // Success will have to be indicated here
+
+                    const bool async_status = spi_transfer_async(&spi_obj, p_tx_buf, BUF_SIZE, p_rx_buf, BUF_SIZE, p_fill_sym, (void*)&some_ctx, dma_modes[mode_id]);
```
This API has changed.

fkjagodzinski reviewed (changes requested):
Also please run `astyle` on files from this PR.

Ok, please ignore my comments regarding the outdated API. I checked existing header instead of the markdown doc.

cmonr commented:
@mprse @fkjagodzinski 
How is #8443 relate to this? Does this still need to be reviewed, or is this considered good to move forward?

cmonr commented:
@mprse @fkjagodzinski

Any updates? If not, then we'll have to close due to inactivity.

mprse commented:
@cmonr Please do not close. This one is needed.

mprse committed:
Define SPI pins for K66F.

Required for testing.

mprse committed:
Define SPI pins for NUCLEO_F429ZI.

This is required for testing.

mprse committed:
Add SPI test and test header file.

This test will run on CI and dose not verify the SPI communication.

mprse commented:
Refreshed the test. 
Reference implementation required by this PR is still not merged, but this one is ready. Please review.

cmonr commented:
@mprse I wonder if this branch needs to be rebased, since it's not passing unittests.

mprse commented:
> @mprse I wonder if this branch needs to be rebased, since it's not passing unittests.

Probably its not passing since there is no new SPI HAL API on the branch.

cmonr commented:
@mprse Ah, gotcha. So this is waiting on https://github.com/ARMmbed/mbed-os/pull/8445?

mprse commented:
Rebased this one after feature branch has been rebased on master.

cmonr commented:
@mprse This needs a rebase.

mprse commented:
This will have to be rebased after PR https://github.com/ARMmbed/mbed-os/pull/8445 is merged.

cmonr commented:
@mprse Rebase away!

mprse committed:
NUCLEO_F429ZI: Fix SPI ref impl - add null check in spi_transfer()

mprse commented:
After #8445 has been merged I have rebased this one and added the following updates:
- cleanup SPI pins config, so it is consistent between SPI tests (this one and one board spi com test),
- Added fix for STM implementation found by the test.

PR is now ready for review and CI.

0xc0170 commented:
I started CI to get some early results

0xc0170 commented:
We are experiencing CI node failures, few jobs were aborted. We are investigating, will restart the jobs once fixed.

cc @ARMmbed/mbed-os-test

mprse commented:
> We are experiencing CI node failures, few jobs were aborted. We are investigating, will restart the jobs once fixed.

Please do not run CI. Looks like the failures are related. I'll check it today.

mprse commented:
PR is now ready for review and CI, but I had problem with specifying SPI pins for testing purposes and I'm not sure if my solution is correct (please see the following commits: 872a66dc962cf6eb303fbfe4b8ac5142905cefac, 9f299f4323a5edfbe855fade7ae8c714691957e3). 

For the testing we need to have two SPI interfaces specified (this test requires at least one, but one board communication test requires two). If board provides only one SPI interface, then com test must be skipped. 
I found the following problems:

1. K66F board does not provide `MOSI`/`MISO` pins instead on this platform we have `SIN`/`SOUT` pins which can be used as `MOSI`/`MISO` depending on the operational mode(master/slave). So in case of SPI master `SIN`  corresponds  `MISO` and `SOUT` to `MOSI`. In case of SPI slave its the opposite, so `SIN` corresponds  `MOSI` and `SOUT` to `MISO`. Because of that it won't be true if we specify for example:
`SPI_0_MOSI   = PTD2` since it will be true only for the master mode. To solve this problem I added a note in `K66F` `PinNames.h` file to inform which SPI pins are designed for master and which for slave:
https://github.com/ARMmbed/mbed-os/blob/872a66dc962cf6eb303fbfe4b8ac5142905cefac/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/TARGET_FRDM/PinNames.h#L236-L254

2. Depending on platform available SPI interfaces are indexed differently. For example on `K66F` we have 
`SPI_0 - SPI_2`. On `NUCLEO_F429ZI` we have `SPI_1 - SPI_6`. This means that I can't always define SPI pins for interfaces 0 and 1 like `SPI_0_MOSI, SPI_0_MISO, SPI_0_CS, SPI_0_SCK, SPI_1_MOSI, SPI_1_MISO, SPI_1_CS, SPI_1_SCK`. And use these pins in the test assuming that SPI_0 is master and SPI_1 is slave. On `NUCLEO_F429ZI` we don't have `SPI_0`. Additionally some SPI interface may only be available for internal usage, e.g. connected to SD card slot. To solve this problem I added additional macros for testing purposes only which specifies which interfaces are designed for testing:
https://github.com/ARMmbed/mbed-os/blob/872a66dc962cf6eb303fbfe4b8ac5142905cefac/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/TARGET_FRDM/PeripheralNames.h#L127-L138

@0xc0170 @donatieng @jamesbeyond Is this acceptable solution?

cmonr commented:
@mprse This PR now needs a rebase.

mprse commented:
Rebased after spi feature branch has been rebased on master to solve Travis issue.

mprse commented:
> fatal: ambiguous argument 'HEAD..feature-hal-spec-spi': unknown revision or path not in the working tree.
> Use '--' to separate paths from revisions, like this:
> 'git <command> [<revision>...] -- [<file>...]'
> The command "git diff --name-only --diff-filter=d HEAD..${TRAVIS_BRANCH} \
>   | ( grep '.\(c\|cpp\|h\|hpp\)$' || true ) \
>   | ( fgrep -v -f .astyleignore || true ) \
>   | while read file; do astyle -n --options=.astylerc "${file}"; done" exited with 128.

@ARMmbed/mbed-os-maintainers 
After rebasing I'm getting above Travis errors. Any ideas? Is that might be related to PR https://github.com/ARMmbed/mbed-os/pull/9289?

0xc0170 commented:
@cmonr

It's not be related to #9289. It is related to this one : https://github.com/ARMmbed/mbed-os/commit/af4af430a2a65e298df0ffc59227eb38a4725bb4#diff-354f30a63fb0907d4ad57269548329e3 . HEad was added, so it's nto checking just the branch but the diff between head and the branch. However it does not known head?

This is a bug

cmonr commented:
Fix is available (https://github.com/ARMmbed/mbed-os/pull/9346), with what appears to be a successful non-master base PR test (https://github.com/ARMmbed/mbed-os/pull/9348).

Two options.
1) Once it's merged, this PR's base feature branch can be rebased and pushed, 
-OR-
2) The `.travis.yml` fix_es_ can be cherry-picked into the feature branch.

This choice is up to the @ARMmbed/mbed-os-maintainers who will be around soon (I would go with #1)

mprse commented:
Target branch has been rebased to solve second Travis issue. This PR has been rebased and it looks like it is ok now!

PR is ready for review and CI.

cmonr commented:
@fkjagodzinski @MarceloSalazar @c1728p9 @ithinuel @screamerbg @jamesbeyond When y'all get a chance.

jamesbeyond commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/TARGET_FRDM/PeripheralNames.h:
```diff
@@ -124,14 +124,19 @@ typedef enum {
+#define SPI_TEST_MASTER SPI_2
+#define SPI_TEST_SLAVE SPI_0
+#define SPI_TEST_MASTER_PIN(SPI_PIN) SPI_2##_##SPI_PIN
+#define SPI_TEST_SLAVE_PIN(SPI_PIN)  SPI_0##_##SPI_PIN
```
I can see the reason you want to adding these macros. I am OK with this.
Russ is working on Pinmaps, @c1728p9 do you have better suggestions?

jamesbeyond reviewed (approved):
Test headers looks good to me, all the rest of changes seems valid.

cmonr commented:
@fkjagodzinski @MarceloSalazar @ithinuel @screamerbg Mind taking a look? It seems discussions is finally settling.

mprse commented:
@0xc0170 Can we run CI here?

fkjagodzinski commented on TESTS/mbed_hal/spi/main.cpp:
```diff
@@ -0,0 +1,560 @@
+
+                core_util_critical_section_exit();
+
+                wait_ms(2);
```
Could you add a comment here and explain the reason for this `wait_ms(2)`?

mprse replied:
fixed.

fkjagodzinski reviewed (approved):
Looks OK! :+1: Only one missing comment.

mprse commented:
I addressed last review comment. Can we run CI?

cmonr commented:
@mprse Before we do, please take a look at the Travis CI failures.

mprse commented:
> @mprse Before we do, please take a look at the Travis CI failures.

I fixed asyle failure. 
But it looks like there are some problems now with Python and Ice tea:
https://travis-ci.org/ARMmbed/mbed-os/jobs/486499767

@0xc0170 It looks like `travis-ci/tools-py2.7` is something new (all checks were green before the last minor/comment fix). Do you maybe know where is the problem?

0xc0170 commented:
CI started

mprse committed:
Disable SPI support for all targets except K66F, NUCLEO_F429ZI

0xc0170 commented:
restarted CI

mprse commented:
Rebasing on master caused that new targets with spi feature enabled appeared. This has been fixed.
There is one bigger issue with `mbed-os-example-sd-driver K64F` we have the following failure:
```
ToolException: ./main.cpp:3:27: fatal error: SDBlockDevice.h: No such file or directory
 #include "SDBlockDevice.h"
```

The K64F board has SPI feature and SD component disabled on the feature branch, so this example should be skipped:

https://github.com/ARMmbed/mbed-os-example-sd-driver/blob/3c3f4e7f9b7c0ffc8c77abebd43e774c7ce2b83e/main.cpp#L1-L14

@theamircohen @ARMmbed/mbed-os-storage Can you help please.

mprse committed:
add DEVICE_SPI guard to TT targets

mprse commented:
Added `DEVICE_SPI` guard to `/targets/TARGET_TT/TARGET_TT_M3HQ/spi_api.c`.

theamirocohen commented:
> Rebasing on master caused that new targets with spi feature enabled appeared. This has been fixed.
> There is one bigger issue with `mbed-os-example-sd-driver K64F` we have the following failure:
> 
> ```
> ToolException: ./main.cpp:3:27: fatal error: SDBlockDevice.h: No such file or directory
>  #include "SDBlockDevice.h"
> ```
> 
> The K64F board has SPI feature and SD component disabled on the feature branch, so this example should be skipped:
> 
> https://github.com/ARMmbed/mbed-os-example-sd-driver/blob/3c3f4e7f9b7c0ffc8c77abebd43e774c7ce2b83e/main.cpp#L1-L14
> 
> @theamircohen @ARMmbed/mbed-os-storage Can you help please.

Hi,
K64F is a component SD board, I believe that part of the example's job is to ensure that component SD isn't mistakenly being disabled from the targets.json file.
If you want to disable it I think a better solution will be to remove the example-sd-driver-example from the examples.json file in this current branch.

mprse committed:
Disable examples unsupported on spi feature branch.

mbed-os-example-sd-driver: on spi feature branch SD componen is disabled for K64F - disable this example.
mbed-os-example-bootloader: example is using SPI pins which are now undefined on NUCLEO_F429ZI - disable this example.

mprse commented:
@theamirocohen Thank you for the solution!

@0xc0170 Can we start CI again?

0xc0170 commented:
CI restarted

0xc0170 commented:
CI restarted (Java error, if repeats, will investigate with CI team)

mprse commented:
Now the results are very strange. In the last commit I modified `examples.json` file as suggested @theamirocohen. I made the list of targets empty for the following examples: `mbed-os-example-sd-driver` and `mbed-os-example-bootloader` as those can't be build on spi feature branch (at least now). So the `mbed-os-example-sd-driver` config looks as follows:

https://github.com/ARMmbed/mbed-os/blob/3d439338d3aa8b363903cb673753fbf133ef63fc/tools/test/examples/examples.json#L356-L368

As a result of these changes (`targets` list is empty) it looks like the examples have been built for all targets, so empty `targets` array means all? @0xc0170 @cmonr  Is that intended behaviour? 

If this is intended behaviour then I believe I should set `compile` flag to `false` for these examples also. Is that correct and sufficient?

theamirocohen commented:
> Now the results are very strange. In the last commit I modified `examples.json` file as suggested @theamirocohen. I made the list of targets empty for the following examples: `mbed-os-example-sd-driver` and `mbed-os-example-bootloader` as those can't be build on spi feature branch (at least now). So the `mbed-os-example-sd-driver` config looks as follows:
> 
> [mbed-os/tools/test/examples/examples.json](https://github.com/ARMmbed/mbed-os/blob/3d439338d3aa8b363903cb673753fbf133ef63fc/tools/test/examples/examples.json#L356-L368)
> 
> Lines 356 to 368 in [3d43933](/ARMmbed/mbed-os/commit/3d439338d3aa8b363903cb673753fbf133ef63fc)
>  { 
>      "name": "mbed-os-example-sd-driver", 
>      "github":"https://github.com/ARMmbed/mbed-os-example-sd-driver", 
>      "mbed": [], 
>      "test-repo-source": "github", 
>      "features" : [], 
>      "targets" : [], 
>      "toolchains" : [], 
>      "exporters": [], 
>      "compile" : true, 
>      "export": true, 
>      "auto-update" : true 
>  } 
> 
> As a result of these changes (`targets` list is empty) it looks like the examples have been built for all targets, so empty `targets` array means all? @0xc0170 @cmonr Is that intended behaviour?
> 
> If this is intended behaviour then I believe I should set `compile` flag to `false` for these examples also. Is that correct and sufficient?

Hi,
I guess that by removing K64F from "targets" you allowed all boards to be built.
My intention was that you completely remove the sd-driver-example.

mprse commented:
Fixed as suggested by @theamirocohen.

@0xc0170 Can we try again?

0xc0170 commented:
CI started

mprse commented:
This one should be ready to go. Later changes (after review) were related to CI problems.

mprse commented:
@donatieng This one should be ready to go. After this one is merged review and CI here is needed:
https://github.com/ARMmbed/mbed-os/pull/8919 .

NirSonnenschein commented:
@ithinuel @MarceloSalazar @screamerbg would you like to review?

donatieng commented:
Thanks a lot @mprse - @ARMmbed/mbed-os-maintainers I'm happy to get this one merged - if @MarceloSalazar and @screamerbg see missing tests we can add them in a following PR

0xc0170 commented on TESTS/mbed_hal/spi/main.cpp:
```diff
@@ -0,0 +1,561 @@
+/*
+ * Copyright (c) 2018 ARM Limited
+ *
```
Can be fixed in separate PR, adding SPDX identifiers for new files

0xc0170 commented:
> I'm happy to get this one merged - if @MarceloSalazar and @screamerbg see missing tests we can add them in a following PR

As this is targeting feature branch, please add your comments when you are able to, they will be addressed. We mark this as ready for merge now to be integrated in the next 24 hours.
